src/plugins/data_source/server/data_source_route_handler_context.ts

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+// eslint-disable-next-line max-classes-per-file
+import { Logger } from 'src/core/server';
+import { IDataSourceClient } from './client/data_source_client';
+
+class OpenSearchDataSourceRouteHandlerContext {
+  constructor(private dataSourceClient: IDataSourceClient, private logger: Logger) {}
+
+  public async getClient(dataSourceId: string) {
+    try {
+      const client = await this.dataSourceClient.asDataSource(dataSourceId);
+      return client;
+    } catch (error) {
+      // TODO: convert as audit log when integrate with osd auditing
+      // https://github.com/opensearch-project/OpenSearch-Dashboards/issues/1986
+      this.logger.error(
+        `Fail to get data source client for dataSource id: [${dataSourceId}]. Detail: ${error.messages}`
+      );
+      throw error;
+    }
+  }
+}
+
+export class DataSourceRouteHandlerContext {
+  readonly opensearch: OpenSearchDataSourceRouteHandlerContext;
```

Comment:
need to handle legacy client as well

Reply:
What's the usage of legacy openesarch client currently? I didn't think too much there as I thought the critical data source consumer,  "search" plugin,  is not relying on legacy clients. Also as I see [here](https://www.elastic.co/guide/en/kibana/7.17/migrating-legacy-plugins.html), the support of legacy is removed after 7.10 for plugins?

> In Kibana 7.10, support for legacy-style Kibana plugins was completely removed. Moving forward, all plugins must be built on the new Kibana Platform Plugin API. This guide is intended to assist plugin authors in migrating their legacy plugin to the Kibana Platform Plugin API.

But I did find some usage through code referencing, still need to dive deep on those. I can create an issue for tracking.
<img width="1176" alt="image" src="https://user-images.githubusercontent.com/32652829/184221282-f6b78691-21fb-4e95-af10-81def3b5b827.png">

Reply:
index pattern service relies on the legacy client to fetch index mappings and fieldCaps, and we have most of our plugins using legacy client as well

Reply:
I see. Created #2133 to track this.
By adding support for legacy client, I assume the route context for data source will look like this
```
datasource: {
  getClient：(id) => Promise<OpenSearchClient>
  legacy: {
    getClient: (id) => Promise<OpenSearchLegacyClient>
  }
}
```